File: app/src/main/java/com/zibete/proyecto1/Adapters/SliderProfileAdapter.java
package com.zibete.proyecto1.Adapters;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.PagerAdapter;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.github.chrisbanes.photoview.PhotoView;
import com.github.clans.fab.FloatingActionButton;
import com.github.clans.fab.FloatingActionMenu;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.zibete.proyecto1.Constants;
import com.zibete.proyecto1.POJOS.Users;
import com.zibete.proyecto1.R;
import com.zibete.proyecto1.SlidePhotoActivity;
import com.zibete.proyecto1.utils.UserRepository;

import java.util.ArrayList;

import static com.zibete.proyecto1.Constants.chatWith;
import static com.zibete.proyecto1.Constants.maxChatSize;
import static com.zibete.proyecto1.utils.FirebaseRefs.ref_datos;

public class SliderProfileAdapter extends PagerAdapter {

    private Context context;
    private LayoutInflater layoutInflater;
    ArrayList <String> photoList;
    PhotoView photo;
    int rotation;

    ImageView ft_perfil;
    ImageView icon_conectado, icon_desconectado;
    TextView nameUser, tv_estado, desc, age, distanceUser;
    FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
    LinearLayout linearImageActivity, linearPhotos, linear_desc;
    RecyclerView recyclerPhotos;
    String unknownName;
    CoordinatorLayout coordinatorLayoutPhoto;
    ArrayList <String> receivedPhotos;
    AdapterPhotoReceived adapterPhotoReceived;
    ImageView perfil_favorite_on, perfil_favorite_off, perfil_bloq, perfil_bloq_me;
    ProgressBar loadingPhoto;
    FloatingActionMenu floatingActionMenu;
    FloatingActionButton subMenu_chatWith, subMenu_chatWithUnknown;
    ArrayList <Users> userList;


    public SliderProfileAdapter(Context context, ArrayList <Users> userList, int rotation){
        this.context = context;
        this.userList = userList;
        this.rotation = rotation;
    }



    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        layoutInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        final View view = layoutInflater.inflate(R.layout.adapter_profile, container, false);

        floatingActionMenu = view.findViewById(R.id.floatingActionMenu);

        subMenu_chatWith = view.findViewById(R.id.subMenu_chatWith);
        subMenu_chatWithUnknown = view.findViewById(R.id.subMenu_chatWithUnknown);

        floatingActionMenu.setClosedOnTouchOutside(true);


        LinearLayoutManager layoutManager = new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, true);
        layoutManager.setStackFromEnd(true);
        recyclerPhotos = view.findViewById(R.id.recyclerPhotos);
        recyclerPhotos.setLayoutManager(layoutManager);

        photoList = new ArrayList<>();
        receivedPhotos = new ArrayList<>();
        adapterPhotoReceived = new AdapterPhotoReceived(receivedPhotos, maxChatSize, context);

        recyclerPhotos.setAdapter(adapterPhotoReceived);

        linearImageActivity = view.findViewById(R.id.linearImageActivity);
        linearPhotos = view.findViewById(R.id.linearPhotos);
        distanceUser = view.findViewById(R.id.distanceUser);
        ft_perfil = view.findViewById(R.id.ftPerfil);
        nameUser = view.findViewById(R.id.nameUser);
        desc = view.findViewById(R.id.desc);
        age = view.findViewById(R.id.edad);
        tv_estado = view.findViewById(R.id.tv_estado);
        icon_conectado = view.findViewById(R.id.icon_conectado);
        icon_desconectado = view.findViewById(R.id.icon_desconectado);

        perfil_favorite_off = view.findViewById(R.id.perfil_favorite_off);
        perfil_favorite_on = view.findViewById(R.id.perfil_favorite_on);
        perfil_bloq = view.findViewById(R.id.perfil_bloq);
        perfil_bloq_me = view.findViewById(R.id.perfil_bloq_me);
        loadingPhoto = view.findViewById(R.id.loadingPhoto);
        coordinatorLayoutPhoto = view.findViewById(R.id.coordinatorLayoutPhoto);
        linear_desc = view.findViewById(R.id.linear_desc);

        final Users users = userList.get(position);

        nameUser.setText(users.getNombre());


        DisplayMetrics metrics = new DisplayMetrics();
        WindowManager windowManager = (WindowManager) context
                .getSystemService(Context.WINDOW_SERVICE);
        windowManager.getDefaultDisplay().getMetrics(metrics);



        int height = metrics.heightPixels;

        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, height-(height/4));

        linearImageActivity.setLayoutParams(layoutParams);



        //FAVORITE LIST
        perfil_favorite_off.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {

                ref_datos.child(user.getUid()).child("FavoriteList").child(users.getID()).setValue(users.getID());

                Toast.makeText(context, "Agregado a favoritos", Toast.LENGTH_SHORT).show();

            }
        });

        perfil_favorite_on.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {

                ref_datos.child(user.getUid()).child("FavoriteList").child(users.getID()).removeValue();

                Toast.makeText(context, "Quitado de favoritos", Toast.LENGTH_SHORT).show();

            }
        });


//Set Data
        if (!users.getDescripcion().isEmpty()) {
            linear_desc.setVisibility(View.VISIBLE);
            desc.setText(users.getDescripcion());
        }else{
            linear_desc.setVisibility(View.GONE);
        }


        final ProgressBar progressBar = new ProgressBar(context);

        CoordinatorLayout.LayoutParams pbParam = new CoordinatorLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        pbParam.gravity = Gravity.CENTER;

        progressBar.setLayoutParams(pbParam);

        coordinatorLayoutPhoto.addView(progressBar);

        Glide.with(context)
                .load(users.getFoto())
                .apply(new RequestOptions().transform( new CenterCrop(), new RoundedCorners(35)))
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .listener(new RequestListener<Drawable>() {
                    @Override
                    public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                        progressBar.setVisibility(View.GONE);
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                        progressBar.setVisibility(View.GONE);
                        return false;
                    }

                })
                .into(ft_perfil);

        

        linearImageActivity.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                photoList.add(users.getFoto());
                Intent intent = new Intent(context, SlidePhotoActivity.class);
                intent.putExtra("photoList",photoList);
                intent.putExtra("position",0);
                intent.putExtra("rotation",180);
                view.getContext().startActivity(intent);
            }
        });


        UserRepository.stateUser(context, users.getID(), icon_conectado, icon_desconectado, tv_estado, chatWith);

        UserRepository.setUserOnline(context, users.getID());


        new Constants().setFavorite(users.getID(), perfil_favorite_on, perfil_favorite_off);

        new Constants().setBloq(users.getID(), perfil_bloq);

        new Constants().getBloqMe(users.getID(), perfil_bloq_me);

        new Constants().getAge(users.getID(), age);

        new Constants().getDistanceToUser(users.getID(), distanceUser);

        new Constants().addPhotoReceived(users.getID(), adapterPhotoReceived, linearPhotos);

        new Constants().setMenuProfile(context, users.getID(), subMenu_chatWithUnknown,subMenu_chatWith);



        container.addView(view);
        return view;
    }




    @Override
    public int getCount() {
        return userList.size();
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view == (CoordinatorLayout) object;
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        container.removeView((CoordinatorLayout)object);
    }




}

File: app/src/main/java/com/zibete/proyecto1/adapters/SliderProfileAdapter.kt
package com.zibete.proyecto1.adapters

import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.util.DisplayMetrics
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.github.clans.fab.FloatingActionButton
import com.github.clans.fab.FloatingActionMenu
import com.google.firebase.auth.FirebaseAuth
import com.zibete.proyecto1.ui.constants.Constants
import com.zibete.proyecto1.R
import com.zibete.proyecto1.SlidePhotoActivity
import com.zibete.proyecto1.model.Users
import com.zibete.proyecto1.utils.FirebaseRefs
import com.zibete.proyecto1.utils.ProfileUiBinder
import com.zibete.proyecto1.data.UserRepository
import javax.inject.Inject

class SliderProfileAdapter @Inject constructor(
    private val profileUiBinder: ProfileUiBinder,
    private val userRepository: UserRepository,
    private val context: Context,
    private val userList: MutableList<Users>,
    private val rotation: Int
) : RecyclerView.Adapter<SliderProfileAdapter.VH>() {

    // --- ViewHolder ---
    inner class VH(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val floatingActionMenu: FloatingActionMenu = itemView.findViewById(R.id.floatingActionMenu)
        val subMenuChatWith: FloatingActionButton = itemView.findViewById(R.id.menu_go_chat)
        val subMenuChatWithUnknown: FloatingActionButton = itemView.findViewById(R.id.menu_go_chat_group)

        val recyclerPhotos: RecyclerView = itemView.findViewById(R.id.recyclerPhotos)
        val linearImageActivity: LinearLayout = itemView.findViewById(R.id.linearImageActivity)
        val linearPhotos: LinearLayout = itemView.findViewById(R.id.linearPhotos)
        val distanceUser: TextView = itemView.findViewById(R.id.distanceUser)
        val ftPerfil: ImageView = itemView.findViewById(R.id.ftPerfil)
        val nameUser: TextView = itemView.findViewById(R.id.nameUser)
        val desc: TextView = itemView.findViewById(R.id.desc)
        val age: TextView = itemView.findViewById(R.id.edad)
        val tvEstado: TextView = itemView.findViewById(R.id.tv_status)
        val iconConectado: ImageView = itemView.findViewById(R.id.icon_connected)
        val iconDesconectado: ImageView = itemView.findViewById(R.id.icon_disconnected)
        val perfilFavoriteOff: ImageView = itemView.findViewById(R.id.perfil_favorite_off)
        val perfilFavoriteOn: ImageView = itemView.findViewById(R.id.perfil_favorite_on)
        val perfilBloq: ImageView = itemView.findViewById(R.id.perfil_bloq)
        val perfilBloqMe: ImageView = itemView.findViewById(R.id.perfil_bloq_me)
        val coordinatorLayoutPhoto: CoordinatorLayout = itemView.findViewById(R.id.coordinatorLayoutPhoto)
    }

    private val currentUser = FirebaseAuth.getInstance().currentUser

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.adapter_profile, parent, false)
        return VH(v)
    }

    override fun getItemCount(): Int = userList.size

    override fun onBindViewHolder(holder: VH, position: Int) {
        val user = userList[position]

        // FAB menu
        holder.floatingActionMenu.setClosedOnTouchOutside(true)

        // Carrusel de “fotos recibidas”
        val receivedPhotos = ArrayList<String>()
        val adapterPhotoReceived = AdapterPhotoReceived(receivedPhotos, Constants.MAXCHATSIZE, context)
        holder.recyclerPhotos.apply {
            val lm = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, true).apply { stackFromEnd = true }
            layoutManager = lm
            adapter = adapterPhotoReceived
        }

        // Nombre
        holder.nameUser.text = user.name

        // Alto dinámico del contenedor de imagen principal (75% pantalla aprox.)
        val metrics: DisplayMetrics = context.resources.displayMetrics
        val height = metrics.heightPixels
        holder.linearImageActivity.layoutParams = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            height - (height / 4)
        )

        // Favoritos ON/OFF
        holder.perfilFavoriteOff.setOnClickListener {
            currentUser?.uid?.let { uid ->
                FirebaseRefs.refDatos.child(uid).child("FavoriteList").child(user.id).setValue(user.id)
                Toast.makeText(context, "Agregado a favoritos", Toast.LENGTH_SHORT).show()
            }
        }
        holder.perfilFavoriteOn.setOnClickListener {
            currentUser?.uid?.let { uid ->
                FirebaseRefs.refDatos.child(uid).child("FavoriteList").child(user.id).removeValue()
                Toast.makeText(context, "Quitado de favoritos", Toast.LENGTH_SHORT).show()
            }
        }

        // Descripción visible/oculta
        if (user.description.isNotEmpty()) {
            holder.desc.text = user.description
            holder.desc.visibility = View.VISIBLE
        } else {
            holder.desc.visibility = View.GONE
        }

        // ProgressBar centrado para la foto de perfil
        val progressBar = ProgressBar(context).apply {
            layoutParams = CoordinatorLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ).also { it.gravity = Gravity.CENTER }
            isIndeterminate = true
        }
        holder.coordinatorLayoutPhoto.addView(progressBar)

        // Foto de perfil (c/ esquinas redondeadas)
        Glide.with(context)
            .load(user.profilePhoto)
            .apply(RequestOptions().transform(CenterCrop(), RoundedCorners(35)))
            .diskCacheStrategy(DiskCacheStrategy.ALL)
            .listener(object : RequestListener<Drawable?> {

                override fun onLoadFailed(
                    e: GlideException?,
                    model: Any?,
                    target: Target<Drawable?>,
                    isFirstResource: Boolean
                ): Boolean {
                    progressBar.visibility = View.GONE
                    return false
                }

                override fun onResourceReady(
                    resource: Drawable?,
                    model: Any?,
                    target: Target<Drawable?>?,
                    dataSource: DataSource?,
                    isFirstResource: Boolean
                ): Boolean {
                    progressBar.visibility = View.GONE
                    return false                }


            })
            .into(holder.ftPerfil)

        // Abrir foto en SlidePhotoActivity (full screen + zoom) — usa la misma lógica que fotos recibidas
        holder.linearImageActivity.setOnClickListener {
            val photoList = arrayListOf(user.profilePhoto)
            val intent = Intent(context, SlidePhotoActivity::class.java).apply {
                putStringArrayListExtra("photoList", photoList)
                putExtra("position", 0)
                putExtra("rotation", 180)
            }
            it.context.startActivity(intent)
        }

        // Estado online/offline + texto
        UserRepository.stateUser(
            context,
            user.id,
            holder.iconConectado,
            holder.iconDesconectado,
            holder.tvEstado,
            Constants.NODE_CURRENT_CHAT
        )
        userRepository.setUserOnline(context, user.id)

        // Binders auxiliares
        profileUiBinder.setFavorite(user.id, holder.perfilFavoriteOn, holder.perfilFavoriteOff)
        profileUiBinder.getBloqMe(user.id, holder.perfilBloqMe)
        profileUiBinder.getAge(user.id, holder.age)

        profileUiBinder.getDistanceToUser(user.id) { distanceText ->
            holder.distanceUser.text = distanceText
        }

        profileUiBinder.addPhotoReceived(user.id, adapterPhotoReceived, holder.linearPhotos)
        profileUiBinder.setMenuProfile(context, user.id, holder.subMenuChatWithUnknown, holder.subMenuChatWith)
    }
}
